### PR TITLE
feat(snapshot-backfill): control log store back pressure in backfill executor

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

In snapshot backfill implemented in https://github.com/risingwavelabs/risingwave/pull/17735, we have a complicated control mechanism on meta node when consuming log store after finishing consuming snapshot. On meta node, snapshot backfill is divided into 4 phases, `ConsumingSnapshot`, `ConsumingLogStore`, `ConsumingUpstream` and `Finishing`. After finishing consuming snapshot, it enters `ConsumingLogStore` phase. All snapshot backfill actors will keep consuming log store until receiving a barrier with special mutation from meta node. After hitting some pre-defined criterion, it will enter `ConsumingUpstream` phase. In this phase, the barrier of upstream and creating job is still decoupled, but the backfill executors have already started consuming upstream. On meta side, we will gradually reduce the number of barriers lagged behind. After the lag is small enough, it enters the `Finishing` phase, after which the creating job will join as part of the upstream, in which barriers are injected and collected together.

The current implementation has some drawbacks. 
* before the barrier to turn `ConsumingLogStore` to `ConsumingUpstream`, backfill executors still consume log store even though it has caught up with the upstream, which unnecessarily still reads log store data even though the upstream data are ready. 
* though we have some control mechanism on meta node to ensure that the number of barrier lagging behind gradually decreases, there is no actual back pressure to upstream on streaming executor side.
* the control state on meta node is too complicated, and can be simplified.

In this PR, as discussed offline with @hzxa21 , we implement a much simpler mechanism for snapshot backfill control. On meta node, there is only 3 phases, `ConsumingSnapshot`, `ConsumingLogStore` and `Finishing`, where the original `ConsumingUpstream` is removed. 

For snapshot backfill executors, after consuming snapshot and entering `ConsumingLogStore`, we will start having back pressure to upstream by limiting the maximum number of barrier lagging behind in the barrier buffer. This maximum number will be decreased gradually to ensure that the phase of `ConsumingLogStore` can be eventually finished. After there is no pending barrier any more, it means it has caught up to the upstream, and then it will start reading data from upstream, and will not read from log store any more. Note that, this process is all controlled by the snapshot backfill executors themselves, and they do not receive any control information from meta node. On the contrary, when they catch up with upstream and start consuming upstream, they will notify meta node about the progress. 

In meta node, when all snapshot backfill executors have caught up with upstream, it will enter the `Finishing` phase directly. This transition happens with a special barrier, which is the same barrier as the previous implementation. On this barrier,
* after this barrier, the creating job will join the upstream, and all later barriers are injected and collected together
* on receiving this barrier, the upstream mv executor will stop writing log store old value, which is the same as the previous implementation.
* snapshot backfill executor will not be aware of this barrier, because it no longer receives control information from meta node after finishing consuming snapshot.
* For this barrier, it is injected and collected independently between upstream and the creating job. However, the upstream and the creating job will commit this epoch together, which means the upstream will wait for the creating job to collect this barrier before committing the epoch. After this epoch is committed, the creating job will be marked as created, and the snapshot backfill is finished.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by kwannoel** [comment]
> 1. snapshot backfill executor will not be aware of this barrier, because it no longer receives control information from meta node after finishing consuming snapshot.
> 2. For this barrier, it is injected and collected independently between upstream and the creating job. However, the upstream and the creating job will commit this epoch together, which means the upstream will wait for the creating job to collect this barrier before committing the epoch. After this epoch is committed, the creating job will be marked as created, and the snapshot backfill is finished.

I'm a little confused by the phrasing here. The first statement states that "snapshot backfill executor will not be aware of this barrier". But in the second statement, it is implied that this barrier will still be "injected and collected from the creating job", so it seems like it will still be received by the snapshot backfill executor. 

Could you elaborate more on what you mean by "not be aware"? I suppose it does not mean "will not receive", but something else entirely. But I'm not quite sure what.

**Comment by wenym1** [comment]
> Could you elaborate more on what you mean by "not be aware"? I suppose it does not mean "will not receive", but something else entirely. But I'm not quite sure what. Do you mean it will just be treated as a normal barrier in the snapshot backfill executor, and not anytihng special? And in contrast, the upstream mv executor will handle it specially, and stop writing to logstore on receiving it?

For `not be aware`, I actually meant not aware of the barrier mutation, which is the key difference to the previous implementation. It only cares about the epoch information in the barrier. In the previous implementation, it depends on the barrier mutation to transit from consuming log store to consuming upstream. Sorry for the confusing wording.

File: src/meta/src/barrier/mod.rs [hunk 1]
```diff
@@ -1389,6 +1353,65 @@ impl CheckpointControl {
...
+                        .map(TableId::new)
+                        .collect();
+                    let join_handle =
+                        tokio::spawn(self.context.clone().complete_creating_job_barrier(
+                            epoch,
+                            resps,
+                            tables_to_commit,
+                            is_first_time,
+                        ));
+                    self.completing_command = CompletingCommand::CreatingStreamingJob {
```
Comment (hzxa21) [inline src/meta/src/barrier/mod.rs:1388]: IIUC, currently we are polling commands for the following barriers here:
1. Collected fake barrier for creating jobs
2. Collected normal barrier for createing jobs
3. Collected normal barrier for created jobs

1 and 2 are biased over 3. Is it possible that this can cause starvation in commit_epoch for 3 and thus delay the regular checkpoint? For example, right before each time `next_completed_barrier` is called, there is a barrier from 1 or 2 finishing its collection. In the worst case, regular ckpt cannot be completed until the new jobs finish creation.

**Review by hzxa21 — COMMENTED** [review]
Left some early comments.

File: src/meta/src/barrier/creating_job/barrier_control.rs [hunk 2]
```diff
@@ -228,46 +175,45 @@ impl CreatingStreamingJobBarrierControl {
...
+    /// Only epoch within the `epoch_end_bound` can be started.
     pub(super) fn start_completing(
         &mut self,
-    ) -> (Vec<u64>, Option<(u64, Vec<BarrierCompleteResponse>, bool)>) {
-        if self.completing_barrier.is_some() {
-            return (vec![], None);
-        }
-        let mut upstream_epochs_to_notify = vec![];
-        while let Some(mut epoch_state) = self.pending_barriers_to_complete.pop_back() {
+        epoch_end_bound: Bound<u64>,
```
Comment (kwannoel) [inline src/meta/src/barrier/creating_job/barrier_control.rs:183]: What is this `epoch_end_bound` referring to, could you add some docs for it?

Comment (kwannoel) [reply src/meta/src/barrier/creating_job/barrier_control.rs:183]: It is clear from the function header, that it has this property:
```
Only epoch within the `epoch_end_bound` can be started.
```

But it's not clear why we need it, what's its function.

File: src/meta/src/barrier/creating_job/status.rs [hunk 3]
```diff
@@ -12,22 +12,91 @@
...
+impl CreateMviewLogStoreProgressTracker {
+    fn new(actors: impl Iterator<Item = ActorId>, initial_pending_count: usize) -> Self {
+        Self {
+            ongoing_actors: HashMap::from_iter(actors.map(|actor| (actor, initial_pending_count))),
+            finished_actors: HashSet::new(),
+        }
+    }
+
+    pub(super) fn gen_ddl_progress(&self) -> String {
+        let sum = self.ongoing_actors.values().sum::<usize>() as f64;
```
Comment (kwannoel) [inline src/meta/src/barrier/creating_job/status.rs:52]: What's this sum, is it the total barrier count we are pending to backfill from the logstore?

File: src/stream/src/executor/backfill/snapshot_backfill.rs [hunk 4]
```diff
@@ -339,61 +323,41 @@ struct StateOfConsumingSnapshot {
...
-            self.barriers.push_front(barrier);
-        }
-        if self.is_finished {
-            assert!(self.upstream_pending_barriers.is_empty());
+impl UpstreamBufferState for StateOfConsumingLogStore {
+    fn can_consume_upstream(&self) -> bool {
+        if let Some(prev_barrier) = self.upstream_pending_barriers.front()
+            && prev_barrier.kind.is_barrier()
+        {
+            // allow consuming upstream when the barrier is non-checkpoint barrier to avoid deadlock
```
Comment (kwannoel) [inline src/stream/src/executor/backfill/snapshot_backfill.rs:356]: Why deadlock can happen if prev_barrier is checkpoint barrier?

File: src/stream/src/executor/backfill/snapshot_backfill.rs [hunk 5]
```diff
@@ -476,59 +443,47 @@ impl<'a, S: UpstreamBufferState> UpstreamBuffer<'a, S> {
...
-            assert!(subscriptions_to_drop.iter().any(
-                |(subscriber_id, subscribed_upstream_table_id)| {
-                    *subscriber_id == current_subscriber_id
-                        && upstream_table_id == *subscribed_upstream_table_id
+impl<'a> UpstreamBuffer<'a, StateOfConsumingLogStore> {
+    async fn next_barrier(&mut self) -> StreamExecutorResult<Option<DispatcherBarrier>> {
+        Ok(
+            if let Some(barrier) = self.state.upstream_pending_barriers.pop_back() {
+                // only update the `max_pending_barrier_num` on checkpoint barrier to avoid deadlock caused by
+                // downstream waiting on a checkpoint barrier to bump up.
```
Comment (kwannoel) [inline src/stream/src/executor/backfill/snapshot_backfill.rs:467]: What do you mean by "downstream waiting on a checkpoint barrier to bump up"? This comment seems to suggest that updating `max_pending_barrier_num` will somehow result in downstream waiting for checkpoint barrier, but I'm not so certain why.

Some comments would be helpful here.

File: src/meta/src/barrier/mod.rs (hunk 1 above)
Comment (wenym1) [reply src/meta/src/barrier/mod.rs:1388]: Will not have starvation after https://github.com/risingwavelabs/risingwave/pull/18819

File: src/meta/src/barrier/creating_job/status.rs (hunk 3 above)
Comment (wenym1) [reply src/meta/src/barrier/creating_job/status.rs:52]: The sum is used to calculate the average number of pending barriers.

File: src/stream/src/executor/backfill/snapshot_backfill.rs (hunk 4 above)
Comment (wenym1) [reply src/stream/src/executor/backfill/snapshot_backfill.rs:356]: Let's say we have epoch 1, 2, 3, where 1, 2 is non-checkpoint epoch, and 3 is checkpoint barrier. After receiving barrier of epoch 1, the snapshot backfill executor will do `iter_log` on epoch 1 to replay the change log, before which it should wait for the committed epoch of upstream to bump to above epoch 1. However, since epoch 1 is a non-checkpoint epoch, we actually need to wait until the upstream to commit epoch 3. But we currently only receive barrier 1 from upstream, and then stop polling, which can have back-pressure on the upstream, and this back pressure may block the upstream from processing epoch 3, and cause the deadlock.

In short, the deadlock will be like, `downstream iter_log on epoch 1` -> `downstream wait on upstream finish epoch 3` -> `upstream blocked by back pressure caused by downstream not polling further barrier` -> `downstream not poll barrier due to being blocking by iter_log on epoch1`

File: src/meta/src/barrier/creating_job/mod.rs [hunk 6]
```diff
@@ -182,52 +158,22 @@ impl CreatingStreamingJobControl {
...
-                Some(max_collected_epoch)
-            }
-        } else {
-            Some(self.backfill_epoch)
+        if matches!(&self.status, CreatingStreamingJobStatus::Finishing(_)) {
+            return None;
         }
+        // TODO: when supporting recoverable snapshot backfill, we should use the max epoch that has committed
+        Some(max(
+            self.barrier_control.max_collected_epoch().unwrap_or(0),
```
Comment (kwannoel) [inline src/meta/src/barrier/creating_job/mod.rs:166]: It's the max between backfill_epoch and max_collected_epoch because of two scenarios:
1. If max_collected_epoch is larger than backfill epoch, means that backfill has already started consuming logstore, and has at least finished reading logstore from max_collected_epoch.
2. If backfill epoch is larger, means that backfill has not started logstore read. We need to pin the backfill epoch, as that's where it will start logstore read from.

Comment (kwannoel) [reply src/meta/src/barrier/creating_job/mod.rs:166]: Is my understanding correct?

File: src/stream/src/task/barrier_manager/progress.rs [hunk 7]
```diff
@@ -174,22 +198,65 @@ impl CreateMviewProgressReporter {
...
+    ) {
+        match self.state {
+            Some(BackfillState::DoneConsumingUpstreamTable(_))
+            | Some(BackfillState::ConsumingLogStore { .. }) => {
+                // valid state
+            }
+            state => {
+                panic!("cannot update log store progress at state {:?}", state)
+            }
+        }
```
Comment (kwannoel) [inline src/stream/src/task/barrier_manager/progress.rs:240]: nit: I think we can just use:
```
if !matches!(BackfillState::DoneConsumingUpstreamTable(_))
            | Some(BackfillState::ConsumingLogStore { .. }) {
    panic!(...)
}
```

File: src/stream/src/task/barrier_manager/progress.rs [hunk 8]
```diff
@@ -174,22 +198,65 @@ impl CreateMviewProgressReporter {
...
+            },
+        );
+    }
+
+    pub(crate) fn finish_consuming_log_store(&mut self, epoch: EpochPair) {
+        match self.state {
+            Some(BackfillState::DoneConsumingUpstreamTable(_))
+            | Some(BackfillState::ConsumingLogStore { .. }) => {
+                // valid state
+            }
```
Comment (kwannoel) [inline src/stream/src/task/barrier_manager/progress.rs:254]: As above.

File: src/stream/src/executor/backfill/snapshot_backfill.rs (hunk 4 above)
Comment (kwannoel) [reply src/stream/src/executor/backfill/snapshot_backfill.rs:356]: Makes sense, thank you

File: src/stream/src/executor/backfill/snapshot_backfill.rs (hunk 5 above)
Comment (kwannoel) [reply src/stream/src/executor/backfill/snapshot_backfill.rs:467]: Same scenario as https://github.com/risingwavelabs/risingwave/pull/18798/files#r1795161729 above.

File: src/meta/src/barrier/creating_job/barrier_control.rs (hunk 2 above)
Comment (wenym1) [reply src/meta/src/barrier/creating_job/barrier_control.rs:185]: Added a comment. 
```
/// Usually `epoch_end_bound` is the upstream committed epoch. This is to ensure that
/// the creating job won't have higher committed epoch than the upstream.
```

File: src/stream/src/executor/backfill/snapshot_backfill.rs (hunk 5 above)
Comment (wenym1) [reply src/stream/src/executor/backfill/snapshot_backfill.rs]: Updated the comment. The reason is the same as https://github.com/risingwavelabs/risingwave/pull/18798#discussion_r1795161729.

File: src/meta/src/barrier/creating_job/mod.rs (hunk 6 above)
Comment (wenym1) [reply src/meta/src/barrier/creating_job/mod.rs:167]: Yes, it's correct.

File: src/meta/src/barrier/mod.rs [hunk 9]
```diff
@@ -997,11 +966,7 @@ impl GlobalBarrierManager {
             .creating_streaming_job_controls
             .values_mut()
         {
-            creating_job.may_inject_fake_barrier(
-                &mut self.control_stream_manager,
-                prev_epoch.value().0,
-                checkpoint,
-            )?
+            creating_job.may_inject_fake_barrier(&mut self.control_stream_manager, checkpoint)?
```
[the remark this answers is not in the document]
Comment (wenym1) [reply src/meta/src/barrier/mod.rs]: Just updated the code.

In the latest code, we won't transit from ConsumingSnapshot to ConsumingLogStore when handling new upstream barrier. Instead, we changed to do the transition in the `collect` method when we see that the snapshot is fully consumed. The previous pending barriers will be injected when we do the transition.

When handling new upstream barrier, we will only inject fake barrier and push the upstream barrier to the buffer, and won't do the transition.

File: src/stream/src/executor/backfill/snapshot_backfill.rs [hunk 10]
```diff
@@ -476,63 +403,54 @@ impl<'a, S: UpstreamBufferState> UpstreamBuffer<'a, S> {
...
-                    *subscriber_id == current_subscriber_id
-                        && upstream_table_id == *subscribed_upstream_table_id
+impl<'a> UpstreamBuffer<'a, ConsumingLogStore> {
+    async fn next_checkpoint_barrier(
+        &mut self,
+    ) -> StreamExecutorResult<Option<Vec<DispatcherBarrier>>> {
+        Ok(
+            if let Some(barriers) = self.upstream_pending_barriers.pop_back() {
+                // sub(1) to ensure that the lag is monotonically decreasing.
+                self.max_pending_checkpoint_barrier_num = min(
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/snapshot_backfill.rs:435]: very elegant implementation

**Review by hzxa21 — APPROVED** [review]
No major comments. LGTM!

**Review by kwannoel — APPROVED** [review]
LGTM

**Comment by xxchan** [comment]
Is it ok to cherry-pick this PR to 2.1? Since my ongoing PR #18925 (I want to cherry-pick it) also changed `BackfillState` 🤪

**Comment by hzxa21** [comment]
@wenym1 I think we can given that this PR doesn't affect the default backill mechnism.
